src/blog/fearless-structured-concurrency.vmd

```diff
@@ -0,0 +1,542 @@
+
+
+```c
+#include <math.h>
+#include <stdio.h>
+#include <pthread.h>
+
+typedef struct {
+  int exponent;
+  int i;
+  int* results;
+} MyThreadArgs;
+
+void *my_thread_main(void* args_raw) {
+  MyThreadArgs* args = (MyThreadArgs*)args_raw;
+  int exponent = args->exponent;
+  int i = args->i;
+  int* results = args->results;
+
+  results[i] = pow(i, exponent); // some expensive calculation
+
+  free(args_raw);
+  return NULL;
+}
+
+int main() {
+  int exponent = 3;
+
+  int results[5];
+
+  // Don't run each iteration one after the other...
+  // run them in parallel, rather than serially.
+  pthread_t threads[5];
+  for (int i = 0; i < 5; i++) {
+    MyThreadArgs* args = (MyThreadArgs*)malloc(sizeof(MyThreadArgs));
+    args->exponent = exponent;
+    args->i = i;
+    args->results = results;
+    pthread_create(&threads[i], NULL, my_thread_main, args);
+  }
+  // Join the threads
+  for (int i = 0; i < 5; i++) {
+    void* returnval = NULL;
+    pthread_join(threads[i], &returnval);
+  }
+
+  for (int i = 0; i < 5; i++) {
+    printf("%d to the %d'th power is %d!\n", i, exponent, results[i]);
+  }
+  return 0;
+}
+```
+
+
+We needed a lot of code to just run `results[i] = pow(i, exponent);` in parallel!
+
+With structured concurrency, we can do that with *just one line*. Let's add a `#pragma omp parallel for` to our original program: [# To run this, you might need to install OpenMP first. For my Mac, I needed to run `brew install libomp` and compile with `clang main.c -fopenmp -L/usr/local/Cellar/libomp/13.0.0/lib`.]
+
+
+```c
+#include <math.h>
+#include <stdio.h>
+#include <omp.h>
+
+int main() {
+  int exponent = 3;
+
+  int results[5];
+  
+  // Launch some threads and run in parallel!
+  #pragma omp parallel for
+  for (int i = 0; i < 5; i++) {
+    results[i] = pow(i, exponent); // some expensive calculation
+  }
+
+  for (int i = 0; i < 5; i++) {
+    printf("%d to the %d'th power is %d!\n", i, exponent, results[i]);
+  }
+  return 0;
+}
+```
+
+
+
+This is *structured concurrency*. It runs our iterations in parallel!
+
+
+Structured concurrency is great for two reasons:
+
+ * It's much easier! Only one line, rather than restructuring our entire program.
+ * The threads can access data from the surrounding scope.
+    * Notice how all of our threads are accessing `results` and `exponent`.
+
+
+Nathaniel Smith makes a great case that we should use structured concurrency rather than directly using thread APIs such as `pthread_create`, `go`, `asyncio.create_task`, etc. [# See [Notes on structured concurrency, or: Go statement considered harmful](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/).] Structured concurrency isn't a silver bullet, of course, but it's definitely a big step forward.
+
+
+However, structured concurrency is still vulnerable to *data races*. Read on to find out what those are, and how another paradigm can fix them. Then, we'll combine the two!
```

Comment:
Since Smith's article mentions Python I'd also recommend going over Python's solution for data races (GIL), and why thats inadequate briefly.

Reply:
it doesnt mention the GIL, and python doesn't actually prevent data races, as defined in this article